### PR TITLE
SSL/TLS-Modules

### PR DESCRIPTION
**Pull request by Captain-T2004**
Added the SSL/TLS modules to scan for ssl certificate vulnerabilities and to scan for old ssl/tls version or bad cipher suite.

#### Checklist
- [x] I have followed the [Contributor Guidelines](https://github.com/OWASP/Nettacker/wiki/Developers#contribution-guidelines).
- [x] The code has been thoroughly tested in my local development environment with flake8 and pylint.
- [x] The code is Python 3 compatible.
- [ ] The code follows the PEP8 styling guidelines with 4 spaces indentation.
- [x] This Pull Request relates to only one issue or only one feature
- [ ] I have referenced the corresponding issue number in my commit message
- [ ] I have added the relevant documentation.
- [ ] My branch is up-to-date with the Upstream master branch.

#### Your development environment
- OS: `Kali GNU/Linux `
- OS Version: `Rolling x`
- Python Version: `Python 3.11.9`

### CODE REVIEW

**Review by arkid15r — COMMENTED** [review]
I've updated the code w/ some refactoring/reformatting.

File: src/nettacker/core/module.py [hunk 1]
```diff
@@ -45,7 +45,13 @@ def __init__(
         self.skip_service_discovery = options.skip_service_discovery
 
         self.discovered_services = None
-        self.ignored_core_modules = ["subdomain_scan", "icmp_scan", "port_scan"]
+        self.ignored_core_modules = [
+            "subdomain_scan",
+            "icmp_scan",
+            "port_scan",
+            "ssl_certificate_scan",
+            "ssl_version_scan",
```
Comment (arkid15r) [inline src/nettacker/core/module.py:53]: Do we really have to extend this list w/ ssl_ modules?

File: tests/core/test_utils.py [hunk 2]
```diff
@@ -52,3 +52,28 @@ def test_select_maximum_cpu_core(self, cpu_count_mock):
                 )
 
             self.assertEqual(utils.select_maximum_cpu_core("invalid"), 1)
+
+    def test_is_weak_hash_algo(self):
+        for algo in ("md2", "md4", "md5", "sha1"):
+            self.assertEqual(utils.is_weak_hash_algo(algo), True)
```
Comment (arkid15r) [inline tests/core/test_utils.py:58]: Here and and everywhere else
```suggestion
            self.assertTrue(utils.is_weak_hash_algo(algo))
```

File: tests/core/test_utils.py [hunk 3]
```diff
@@ -52,3 +52,28 @@ def test_select_maximum_cpu_core(self, cpu_count_mock):
...
+        self.assertEqual(utils.is_weak_hash_algo("test_aglo"), False)
```
Comment (arkid15r) [inline tests/core/test_utils.py:59]:
```suggestion
        self.assertFalse(utils.is_weak_hash_algo("test_aglo"))
```

File: tests/core/test_socket.py [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+        return self.expired
+
+    def get_notAfter(self):
+        return self.expire_date
+
+    def get_signature_algorithm(self):
+        return self.signature_algorithm
+
+
+class TestSSLMethods(TestCase):
```
Comment (arkid15r) [inline tests/core/test_socket.py:44]: Could you group/split the test cases by good/bad cert input?

File: src/nettacker/core/utils/common.py [hunk 5]
```diff
@@ -324,3 +326,40 @@ def sort_dictionary(dictionary):
...
+            socket_connection = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+            socket_connection.settimeout(timeout)
+            socket_connection.connect((host, port))
+        except ConnectionRefusedError:
+            continue
+        try:
+            socket_connection = ssl.wrap_socket(socket_connection, ciphers=ciphers)
+            bad_cipher_suite = True
+            break
+        except Exception:
```
Comment (arkid15r) [inline src/nettacker/core/utils/common.py:363]: Is it possible to narrow down the expected Exception to minimize false negatives?

File: src/nettacker/core/utils/common.py [hunk 6]
```diff
@@ -324,3 +326,40 @@ def sort_dictionary(dictionary):
...
+        if unsafe_algo in algo:
+            return True
+    return False
+
+
+def check_ssl_version(ssl_ver):
+    return ssl_ver not in {"TLSv1.2", "TLSv1.3"}
+
+
+def check_cipher_suite(host, port, timeout):
```
Comment (arkid15r) [inline src/nettacker/core/utils/common.py:343]: Would it be more consistent if you used `is_weak_ssl_version` and `is_weak_cipher_suite` names?

File: tests/core/test_utils.py [hunk 7]
```diff
@@ -52,3 +52,28 @@ def test_select_maximum_cpu_core(self, cpu_count_mock):
...
+
+    def test_check_ssl_version(self):
+        for ver in {"TLSv1.2", "TLSv1.3"}:
+            self.assertEqual(utils.check_ssl_version(ver), False)
+        self.assertEqual(utils.check_ssl_version("test_version"), True)
+
+    @patch("socket.socket")
+    @patch("ssl.wrap_socket")
+    def test_check_cipher_suite(self, mock_wrap, mock_socket):
```
Comment (arkid15r) [inline tests/core/test_utils.py:68]: It seems we lack tests based on the cipher value here

**Review by arkid15r — CHANGES_REQUESTED** [review]
Looks great, please take a look into addressing the comments when you get a chance:

**Comment by arkid15r** [comment]
@securestep9 I'm unable to add you as a reviewer here. Could you review when you get a chance?

File: src/nettacker/core/module.py (hunk 1 above)
Comment (Captain-T2004) [reply src/nettacker/core/module.py:53]: I have tried to run it without populating this list and it return: "KeyError: 'socket'", this happens with all the modules using socket library. I can look into it more but without any changes somewhere in the code the modules won't work without populating this list.

**Comment by Captain-T2004** [comment]
@arkid15r, I have made the requested changes please review it again. Also on your last comment, I can't think of how to add cipher value based test as per the current login if the ssl connection is established using the provided unsecure_cipher than the method returns True else it returns False. So how can I write a test based on the cipher value?

**Review by arkid15r — APPROVED** [review]
No worries, I misread the function code. The tests are fine.

LGTM

**Pull request merged**